Improved API call loading/failure handling

gilluminate opened the pull request:
Closes [ENG-940]

### Description Of Changes

1. Improved compliance by automatically defaulting any property accessed on `Fides.consent` as `false` in case an experience fails to load or is empty. (ie. `Fides.consent.whatever` will return `false`)

2. Improved TCF (Transparency and Consent Framework) error handling and user experience by implementing retry mechanisms for API calls and better loading states. When the full TCF experience fails to load due to network issues or API failures, users now see appropriate loading indicators and error messages, while still being able to accept/reject consent. The system now uses a retry-able fetch hook with exponential back-off for more reliable API communication.

### Code Changes

* Added `useRetryableFetch` hook to `TcfOverlay` with retry logic for fetching full TCF experience
* Created `TcfLoadingErrorMessage` component to display user-friendly error messages when TCF data fails to load
* Added `RecordsListSkeletons` component for skeleton loading states during data fetching
* Removed manual loading state management from `ConsentButtons` that was causing UX issues
* Updated CSS variables and styles to support skeleton loading animations and error states without affecting existing customer styles.
* Enhanced `InfoBox` component to accept additional props for error styling
* Fixed accessibility issues by adding `tabIndex` to scrollable content and improving `aria-checked` values
* Increased TCF bundle size limit from 92KB to 95KB to accommodate new features (increased by 0.4k, now totaling 92.46k)
* Simplified consent button logic by removing `isMinimalTCF` prop dependencies
* Updated type definitions to allow `false` return values from render functions for when we remove the save button on failure.

### Steps to Confirm
1. Navigate to the demo page in the Vercel deploy with TCF consent banner (`/fides-js-demo.html?geolocation=eea`)
2. Block network requests to simulate API failures and verify error messages appear
3. Test that users can still accept/reject consent even when full experience fails to load
4. Verify skeleton loading states appear during data fetching
5. Confirm retry mechanism works by restoring network and observing successful data load

<img width="1526" height="802" alt="CleanShot 2025-07-25 at 14 10 45@2x" src="https://github.com/user-attachments/assets/6f338e60-eb92-49e0-b8fb-b9feef149653" />

<img width="1444" height="820" alt="CleanShot 2025-07-25 at 14 10 29@2x" src="https://github.com/user-attachments/assets/6e6b5a33-77da-480d-bf65-0cc503ef07e0" />

### Pre-Merge Checklist

* [x] Issue requirements met
* [x] All CI pipelines succeeded
* [x] `CHANGELOG.md` updated
  * [ ] Add a https://github.com/ethyca/fides/labels/db-migration label to the entry if your change includes a DB migration
  * [ ] Add a https://github.com/ethyca/fides/labels/high-risk label to the entry if your change includes a high-risk change (i.e. potential for performance impact or unexpected regression) that should be flagged
  * [ ] Updates unreleased work already in Changelog, no new entry necessary
* Followup issues:
  * [ ] Followup issues created
  * [x] No followup issues
* Database migrations:
  * [ ] Ensure that your downrev is up to date with the latest revision on `main`
  * [ ] Ensure that your `downgrade()` migration is correct and works
    * [ ] If a downgrade migration is not possible for this change, please call this out in the PR description!
  * [x] No migrations
* Documentation:
  * [ ] Documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] Documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
  * [ ] If there are any new client scopes created as part of the pull request, remember to update public-facing documentation that references our scope registry
  * [x] No documentation updates required


[ENG-940]: https://ethyca.atlassian.net/browse/ENG-940?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

tvandort commented on clients/fides-js/src/lib/hooks/useRetryableFetch.ts:
```diff
@@ -0,0 +1,107 @@
+  enabled = true,
+}: UseRetryableFetchOptions<T>): UseRetryableFetchResult<T> => {
+  const [data, setData] = useState<T | undefined>(undefined);
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState(false);
```
Might be better to combine these into a `fetchState: 'loading' | 'error' | 'success'` too. It seems to me that loading and error are mutually exclusive and this would communicate that relationship through the type system instead of through code.

tvandort replied:
Tanstack Query internally represents the state of the query the way I'm recommending:
https://github.com/TanStack/query/blob/main/packages/query-core/src/types.ts#L618
https://github.com/TanStack/query/blob/main/packages/query-core/src/query.ts#L580

Managing a state-machine is much easier long term than managing a number of boolean states. 

I'm less familiar with RTK but it looks like the active state is also represented via an enumeration there: https://github.com/reduxjs/redux-toolkit/blob/master/packages/toolkit/src/query/core/apiState.ts#L71

I do accept that both libraries provide properties that allow library users to quickly check `.isLoading` or `.isErrored` but they do not manage their internal state that way. I think it would be reasonable for us to keep an `isLoading` and `isError` as part of the output interface here but I'd generally say we should make it a practice to not have that representation of the data internally.

gilluminate replied:
This is what I'm referring to https://tanstack.com/query/v5/docs/framework/react/reference/useQuery

gilluminate replied:
updated!

tvandort commented on clients/fides-js/src/lib/initialize.ts:
```diff
@@ -408,14 +410,37 @@ export const initialize = async ({
+    !!fides.experience?.non_applicable_privacy_notices ||
+    !!fides.experience?.privacy_notices;
+
+  // Any unknown consent values should return falsy values automatically
```
Checking my understanding:

1. When an unknown consent flag is accessed we return `false`.
2. If an experience does not load that means all consent flags are "unknown".
3. Therefore, when experiences fail to load, this code is in play "by default" to say that the user has not given consent.

tvandort commented on clients/fides-js/src/lib/hooks/useRetryableFetch.ts:
```diff
@@ -0,0 +1,107 @@
+
+  const finalConfig = { ...DEFAULT_CONFIG, ...config };
+
+  const executeWithRetry = async (attempt = 1): Promise<void> => {
```
Because retry functions can be fiddly I think we should add a unit test around this.

lucanovera reviewed (approved):
I tested the feature locally, retry works correctly with exponential backoff, error message displays, you can still opt in/out to all. Code changes look good. Approved!
Two minor comments:
- Can we remove the cursor pointer from the skeleton elements? (I could not figure out where that property is coming from lol)
- There's a TODO regarding translations, is there a follow up ticket for that?

tvandort reviewed (commented):
This is great, thank you for all these improvements!

A couple of thoughts I had: 

- Should the cursor become a pointer over the skeleton items?
- Would we be able to leave the save button but disable it instead of hiding it when the experience fails to load? I'm worried the buttons expanding might cause mis-taps

tvandort commented on clients/fides-js/src/components/tcf/TcfOverlay.tsx:
```diff
@@ -199,6 +200,38 @@ export const TcfOverlay = () => {
+            ...result,
+            ...userPrefs,
+          };
+          window.Fides.experience = {
+            ...window.Fides.experience,
+            ...fullExperience,
+          };
+          window.Fides.experience.minimal_tcf = false;
```
Ah yeah... definitely. I guess since we're in a world where we don't get it from the backend what I was looking for is a safer pattern for updating a global variable. There might not be much of a better choice here however.

gilluminate replied:
taking another look at this. Yes, the code here was copied from a non-hook source and this code likely doesn't belong in the hook. I'll update that. Thanks.

gilluminate replied:
This is happening as a result of the hook, not in the hook. Maybe I'm not understanding what you are getting at. We get the full tcf data back, and we update the global experience that's already in window.Fides with the full experience.

gilluminate replied:
looks like we replied over top of each other. 👍 On the same page though

gilluminate commented:
@tvandort I've addressed your feedback, please review again?
@lucanovera & @tvandort I've remove the hover states from the skeletons and the save button is now retained but as disabled.

tvandort commented on clients/fides-js/src/components/tcf/TcfOverlay.tsx:
```diff
@@ -680,63 +696,63 @@ export const TcfOverlay = () => {
+              onClose();
+            }}
+            renderFirstButton={() => (
+              <Button
```
<img width="1394" height="403" alt="image" src="https://github.com/user-attachments/assets/ec74e449-568e-48b4-b250-f8c594295d38" />

Should the button outline still be black? IMO this makes the button look "focused". Is there a gray we can use here instead?

gilluminate replied:
This is the "Primary" color, and in our demo's case the primary color is set to our Minos color. It won't necessarily be this color for everyone. (Cookie house is some kind of red currently)

gilluminate replied:
Oh, sorry. I'm trying to move to fast. I missed the word "outline" and thought you were talking about the black buttons... Yeah, I like your suggestion. I'll update the disabled styling.

tvandort commented on clients/fides-js/src/lib/initialize.ts:
```diff
@@ -408,14 +410,37 @@ export const initialize = async ({
     options.fidesConsentNonApplicableFlagMode ?? undefined,
   );
 
+  const hasPrivacyNotices =
```
After any consent action (save, reject, accept) the consent object goes back to being a POJO. Is that expected?

gilluminate replied:
I moved the Proxy logic to be part of the `applyOverridesToConsent` util to avoid this issue. We need to run consent through that every time already anyway.

gilluminate replied:
That wasn't the correct change. `applyOverridesToConsent` is used in places where this causes a problem. I reverted that change and created a new `createConsentProxy` util to use only when we need it.
